### PR TITLE
[LT] Trace through DDP

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77940
* __->__ #77813
* #77806
* #77686
* #77670
* #77443
* #77408
* #76946
* #76722

Summary:
This is PoC of how LazyTensor can trace through DDP. It produces correct
results with limitations stated in the test case.

Test Plan:
python test/lazy/test_ddp.py

### CODE REVIEW

File: test/lazy/test_ddp.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    output = model(input)
+
+    loss = loss_fn(output, labels)
+    loss.backward()
+    optimizer.step()
+    # FIXME(alanwaketan): Investigate why we produce different
+    # results if mark_step.
+    # if device.type == "lazy":
+    #     torch._lazy.mark_step(str(device))
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply test/lazy/test_ddp.py:63]: If I understand correctly the issue here is the `view` relationship will be break after `mark_step`. However if this is the step end I think that should be fine? We generally don't expect tensors' life time exceed the `mark_step`.

[the remark this answers is not in the document]
Comment (wconstab) [reply test/lazy/test_ddp.py:63]: it looks like the mark_step here is the only place you do a mark step.

if you do not do a mark step, then you rely on calling .cpu() when you pass loss to allclose, and then another .cpu() when you pass model parameters to allclose.

in theory calling mark step first should only make those two .cpu() more predictable, they should both return 'already cached' results.
* TODO: can you confirm that when mark_step is used, both .cpu() calls are actually grabbing cached data?

without calling mark_step, i guess each .cpu() call below would re-execute things.  It seems like that should still be correct.

I think you also said there are still graph-splits in this program.  It would probably be worth simplifying things until you can get it into one graph, just for simpler debugging.

[the remark this answers is not in the document]
Comment (wconstab) [reply test/lazy/test_ddp.py:63]: it looks like the mark_step here is the only place you do a mark step.

if you do not do a mark step, then you rely on calling .cpu() when you pass loss to allclose, and then another .cpu() when you pass model parameters to allclose.

in theory calling mark step first should only make those two .cpu() more predictable, they should both return 'already cached' results.
* TODO: can you confirm that when mark_step is used, both .cpu() calls are actually grabbing cached data?

without calling mark_step, i guess each .cpu() call below would re-execute things.  It seems like that should still be correct.

I think you also said there are still graph-splits in this program.  It would probably be worth simplifying things until you can get it into one graph, just for simpler debugging.

**Comment by alanwaketan** [comment]
> @alanwaketan are you still working on this?
> 
> cc @wconstab @Krovatkin

I just transferred to the PyTorch/XLA team at Google. So, I probably will not work on the TorchScript backend anymore. However, the PyTorch/XLA team is still interested at exploring this tracing through DDP idea.

**Comment by byronyi** [comment]
> > @alanwaketan are you still working on this?
> > cc @wconstab @Krovatkin
> 
> I just transferred to the PyTorch/XLA team at Google. So, I probably will not work on the TorchScript backend anymore. However, the PyTorch/XLA team is still interested at exploring this tracing through DDP idea.

That sounds great. 

Btw, does the torch-xla team have anything specific in mind for what is next, given that we have a torch.distributed backend for XLA from https://github.com/pytorch/xla/pull/3339?

**Comment by JackCaoG** [comment]
@byronyi We are prioritizing to make torch.distributed backend work first, we observed some issue when using that backend with new PJRT TPU backend.
